src/persist-client/src/internal/state.rs

```diff
@@ -479,47 +479,14 @@ where
         Continue(state)
     }
 
-    pub fn register_writer(
-        &mut self,
-        hostname: &str,
-        writer_id: &WriterId,
-        purpose: &str,
-        lease_duration: Duration,
-        heartbeat_timestamp_ms: u64,
-    ) -> ControlFlow<NoOpStateTransition<(Upper<T>, WriterState<T>)>, (Upper<T>, WriterState<T>)>
-    {
-        let upper = Upper(self.trace.upper().clone());
-        let writer_state = WriterState {
-            debug: HandleDebugState {
-                hostname: hostname.to_owned(),
-                purpose: purpose.to_owned(),
-            },
-            last_heartbeat_timestamp_ms: heartbeat_timestamp_ms,
-            lease_duration_ms: u64::try_from(lease_duration.as_millis())
-                .expect("lease duration as millis must fit within u64"),
-            most_recent_write_token: IdempotencyToken::SENTINEL,
-            most_recent_write_upper: Antichain::from_elem(T::minimum()),
-        };
-
-        // If the shard-global upper and since are both the empty antichain,
-        // then no further writes can ever commit and no further reads can be
-        // served. Optimize this by no-op-ing writer registration so that we can
-        // settle the shard into a final unchanging tombstone state.
-        if self.is_tombstone() {
-            return Break(NoOpStateTransition((upper, writer_state)));
-        }
-
-        // TODO: Handle if the reader or writer already exists.
-        self.writers.insert(writer_id.clone(), writer_state.clone());
-        Continue((Upper(self.trace.upper().clone()), writer_state))
-    }
-
     pub fn compare_and_append(
         &mut self,
         batch: &HollowBatch<T>,
         writer_id: &WriterId,
         heartbeat_timestamp_ms: u64,
+        lease_duration_ms: u64,
         idempotency_token: &IdempotencyToken,
+        debug_info: &HandleDebugState,
```

Comment:
TODO: pull the writer id, lease duration, and debug info into some `WriterConfig` struct that we can thread around.